File: app/src/main/java/com/apps/game/GamePasswordDialog.java
package com.apps.game;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.HapticFeedbackConstants;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.apps.theme.LauncherMotion;
import com.apps.theme.LauncherTheme;
import com.yuki.yukihub.R;

import java.security.MessageDigest;

/**
 * 九宫格数字密码弹窗，支持设置密码（两次确认）和验证密码两种模式。
 * 密码为 6 位纯数字，存储时使用 SHA-256 哈希。
 */
public final class GamePasswordDialog {

    public interface OnPasswordSetListener {
        void onPasswordSet(String hashedPassword);
    }

    private static final int PASSWORD_LENGTH = 6;
    private static final int MODE_SET = 0;
    private static final int MODE_VERIFY = 1;

    private GamePasswordDialog() {
    }

    /** 设置密码弹窗：输入一次 → 再次确认 → 一致则回调 */
    public static void showSetDialog(Context context, String gameTitle, OnPasswordSetListener listener) {
        show(context, gameTitle, MODE_SET, null, listener, null);
    }

    /** 验证密码弹窗：输入一次 → 与 hashedPassword 比对 → 正确则回调 */
    public static void showVerifyDialog(Context context, String gameTitle, String hashedPassword, Runnable onSuccess) {
        show(context, gameTitle, MODE_VERIFY, hashedPassword, null, onSuccess);
    }

    private static void show(Context context, String gameTitle, int mode,
                             String hashedPassword, OnPasswordSetListener setListener, Runnable verifySuccess) {
        Dialog dialog = new Dialog(context);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

        boolean landscape = context.getResources().getDisplayMetrics().widthPixels
                > context.getResources().getDisplayMetrics().heightPixels;
        int padH = dp(context, landscape ? 20 : 24);
        int padV = dp(context, landscape ? 16 : 28);
        int btnSize = dp(context, landscape ? 42 : 56);
        int btnSpacing = dp(context, landscape ? 4 : 6);
        int sectionGap = dp(context, landscape ? 8 : 16);
        int keypadTop = dp(context, landscape ? 12 : 24);

        LinearLayout root = new LinearLayout(context);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER_HORIZONTAL);
        root.setPadding(padH, padV, padH, padV);
        root.setBackgroundResource(R.drawable.launcher_dialog_bg);

        // 标题
        TextView title = new TextView(context);
        title.setText(mode == MODE_SET ? "设置密码" : "输入密码");
        title.setGravity(Gravity.CENTER);
        title.setTextColor(LauncherTheme.text(context));
        title.setTextSize(16);
        title.setTypeface(null, Typeface.BOLD);
        root.addView(title);

        // 副标题（游戏名）
        if (gameTitle != null && !gameTitle.trim().isEmpty()) {
            TextView subtitle = new TextView(context);
            subtitle.setText(gameTitle);
            subtitle.setGravity(Gravity.CENTER);
            subtitle.setSingleLine(true);
            subtitle.setEllipsize(android.text.TextUtils.TruncateAt.END);
            subtitle.setTextColor(LauncherTheme.textMuted(context));
            subtitle.setTextSize(12);
            LinearLayout.LayoutParams subLp = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            subLp.topMargin = dp(context, 4);
            root.addView(subtitle, subLp);
        }

        // 提示文字
        TextView hint = new TextView(context);
        hint.setText(mode == MODE_SET ? "请输入 6 位数字密码" : "请输入密码以启动游戏");
        hint.setGravity(Gravity.CENTER);
        hint.setTextColor(LauncherTheme.textMuted(context));
        hint.setTextSize(11);
        LinearLayout.LayoutParams hintLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        hintLp.topMargin = dp(context, landscape ? 8 : 12);
        root.addView(hint, hintLp);

        // 密码圆点指示器
        final View[] dots = new View[PASSWORD_LENGTH];
        LinearLayout dotsRow = new LinearLayout(context);
        dotsRow.setOrientation(LinearLayout.HORIZONTAL);
        dotsRow.setGravity(Gravity.CENTER);
        LinearLayout.LayoutParams dotsRowLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        dotsRowLp.topMargin = sectionGap;
        int dotSize = dp(context, 12);
        int dotSpacing = dp(context, 10);
        for (int i = 0; i < PASSWORD_LENGTH; i++) {
            View dot = new View(context);
            GradientDrawable dotBg = new GradientDrawable();
            dotBg.setShape(GradientDrawable.OVAL);
            dotBg.setColor(LauncherTheme.card(context));
            dot.setBackground(dotBg);
            LinearLayout.LayoutParams dotLp = new LinearLayout.LayoutParams(dotSize, dotSize);
            if (i > 0) dotLp.leftMargin = dotSpacing;
            dotsRow.addView(dot, dotLp);
            dots[i] = dot;
        }
        root.addView(dotsRow, dotsRowLp);

        // 九宫格数字键盘
        final StringBuilder input = new StringBuilder();
        final boolean[] firstInputDone = {false};
        final String[] firstInput = {""};

        LinearLayout keypad = new LinearLayout(context);
        keypad.setOrientation(LinearLayout.VERTICAL);
        keypad.setGravity(Gravity.CENTER_HORIZONTAL);
        LinearLayout.LayoutParams keypadLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        keypadLp.topMargin = keypadTop;

        String[] keys = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "", "0", "del"};

        for (int row = 0; row < 4; row++) {
            LinearLayout rowLayout = new LinearLayout(context);
            rowLayout.setOrientation(LinearLayout.HORIZONTAL);
            rowLayout.setGravity(Gravity.CENTER);
            for (int col = 0; col < 3; col++) {
                int idx = row * 3 + col;
                String key = keys[idx];
                if (key.isEmpty()) {
                    View spacer = new View(context);
                    LinearLayout.LayoutParams spacerLp = new LinearLayout.LayoutParams(btnSize, btnSize);
                    if (col > 0) spacerLp.leftMargin = btnSpacing;
                    rowLayout.addView(spacer, spacerLp);
                    continue;
                }
                TextView btn = new TextView(context);
                btn.setText(key.equals("del") ? "X" : key);
                btn.setGravity(Gravity.CENTER);
                btn.setTextSize(20);
                btn.setTypeface(null, Typeface.NORMAL);
                btn.setTextColor(LauncherTheme.primary(context));
                LinearLayout.LayoutParams btnLp = new LinearLayout.LayoutParams(btnSize, btnSize);
                if (col > 0) btnLp.leftMargin = btnSpacing;
                rowLayout.addView(btn, btnLp);

                btn.setOnClickListener(v -> {
                    v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY);
                    if (key.equals("del")) {
                        if (input.length() > 0) {
                            input.deleteCharAt(input.length() - 1);
                        }
                    } else {
                        if (input.length() < PASSWORD_LENGTH) {
                            input.append(key);
                        }
                    }
                    updateDots(dots, input.length(), context);

                    if (input.length() == PASSWORD_LENGTH) {
                        String entered = input.toString();
                        input.setLength(0);
                        // 延迟清空圆点，让用户看到最后一个点亮起
                        dots[0].postDelayed(() -> {
                            updateDots(dots, 0, context);
                            handleComplete(context, dialog, mode, entered, hashedPassword,
                                    firstInputDone, firstInput, hint, setListener, verifySuccess);
                        }, 100);
                    }
                });
            }
            LinearLayout.LayoutParams rowLp = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            if (row > 0) rowLp.topMargin = btnSpacing;
            keypad.addView(rowLayout, rowLp);
        }
        root.addView(keypad, keypadLp);

        // 取消按钮
        TextView cancelBtn = new TextView(context);
        cancelBtn.setText("取消");
        cancelBtn.setGravity(Gravity.CENTER);
        cancelBtn.setTextSize(13);
        cancelBtn.setTypeface(null, Typeface.BOLD);
        cancelBtn.setTextColor(LauncherTheme.primary(context));
        LinearLayout.LayoutParams cancelLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, dp(context, 38));
        cancelLp.topMargin = dp(context, landscape ? 12 : 20);
        cancelBtn.setBackground(LauncherTheme.cancelChip(context));
        cancelBtn.setOnClickListener(v -> dialog.dismiss());
        root.addView(cancelBtn, cancelLp);

        ScrollView scroll = new ScrollView(context);
        scroll.addView(root);
        dialog.setContentView(scroll);
        dialog.show();
        LauncherMotion.applyDialogMotion(dialog);
        Window window = dialog.getWindow();
        if (window != null) {
            window.setBackgroundDrawableResource(android.R.color.transparent);
            int dialogHeight;
            if (landscape) {
                int screenHeight = context.getResources().getDisplayMetrics().heightPixels;
                dialogHeight = (int) (screenHeight * 0.9f);
            } else {
                dialogHeight = WindowManager.LayoutParams.WRAP_CONTENT;
            }
            window.setLayout(dp(context, 280), dialogHeight);
        }
    }

    private static void handleComplete(Context context, Dialog dialog, int mode,
                                       String entered, String hashedPassword,
                                       boolean[] firstInputDone, String[] firstInput,
                                       TextView hint, OnPasswordSetListener setListener,
                                       Runnable verifySuccess) {
        if (mode == MODE_SET) {
            if (!firstInputDone[0]) {
                firstInput[0] = entered;
                firstInputDone[0] = true;
                hint.setText("请再次输入以确认");
            } else {
                if (entered.equals(firstInput[0])) {
                    String hashed = hash(entered);
                    dialog.dismiss();
                    if (setListener != null) setListener.onPasswordSet(hashed);
                } else {
                    firstInputDone[0] = false;
                    firstInput[0] = "";
                    hint.setText("两次输入不一致，请重新设置");
                    shakeError(hint);
                }
            }
        } else {
            String hashed = hash(entered);
            if (hashed.equals(hashedPassword)) {
                dialog.dismiss();
                if (verifySuccess != null) verifySuccess.run();
            } else {
                hint.setText("密码错误，请重新输入");
                shakeError(hint);
                Toast.makeText(context, "密码错误", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private static void updateDots(View[] dots, int count, Context context) {
        int activeColor = LauncherTheme.primary(context);
        int inactiveColor = LauncherTheme.card(context);
        for (int i = 0; i < dots.length; i++) {
            GradientDrawable bg = new GradientDrawable();
            bg.setShape(GradientDrawable.OVAL);
            if (i < count) {
                bg.setColor(activeColor);
            } else {
                bg.setColor(inactiveColor);
            }
            dots[i].setBackground(bg);
        }
    }

    private static void shakeError(View view) {
        view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
    }

    /** SHA-256 哈希，结果转 hex 字符串 */
    public static String hash(String raw) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] bytes = md.digest(raw.getBytes("UTF-8"));
            StringBuilder sb = new StringBuilder();
            for (byte b : bytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (Exception e) {
            return raw;
        }
    }

    private static int dp(Context context, int value) {
        return (int) (value * context.getResources().getDisplayMetrics().density + 0.5f);
    }
}

File: app/src/main/java/com/apps/game/GamePasswordDialog.kt
package com.apps.game

import android.app.Dialog
import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.util.Log
import android.view.Gravity
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import com.apps.theme.LauncherDialogFactory
import com.apps.theme.LauncherMotion
import com.apps.theme.LauncherTheme
import com.core.R
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * 九宫格数字密码弹窗，支持设置密码（两次确认）和验证密码两种模式。
 * 密码为 6 位纯数字，存储时使用 SHA-256 哈希。
 */
object GamePasswordDialog {

    private const val TAG = "GamePasswordDialog"
    private const val PASSWORD_LENGTH = 6
    private const val MODE_SET = 0
    private const val MODE_VERIFY = 1

    fun interface OnPasswordSetListener {
        fun onPasswordSet(hashedPassword: String)
    }

    /** 设置密码弹窗：输入一次 → 再次确认 → 一致则回调 */
    @JvmStatic
    fun showSetDialog(context: Context, gameTitle: String, listener: OnPasswordSetListener) {
        show(context, gameTitle, MODE_SET, null, listener, null)
    }

    /** 验证密码弹窗：输入一次 → 与 hashedPassword 比对 → 正确则回调 */
    @JvmStatic
    fun showVerifyDialog(context: Context, gameTitle: String, hashedPassword: String?, onSuccess: Runnable?) {
        show(context, gameTitle, MODE_VERIFY, hashedPassword, null, onSuccess)
    }

    /** SHA-256 哈希，结果转 hex 字符串。失败时抛出异常，禁止返回明文以避免明文落库。 */
    @JvmStatic
    fun hash(raw: String): String {
        try {
            val md = MessageDigest.getInstance("SHA-256")
            val bytes = md.digest(raw.toByteArray(Charsets.UTF_8))
            val sb = StringBuilder()
            for (b in bytes) {
                sb.append(String.format("%02x", b))
            }
            return sb.toString()
        } catch (e: NoSuchAlgorithmException) {
            Log.e(TAG, "hash failed, refusing to return raw password", e)
            throw IllegalStateException("Failed to encrypt password", e)
        }
    }

    private fun show(context: Context, gameTitle: String, mode: Int,
                     hashedPassword: String?, setListener: OnPasswordSetListener?, verifySuccess: Runnable?) {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)

        val landscape = context.resources.displayMetrics.widthPixels > context.resources.displayMetrics.heightPixels
        val padH = LauncherTheme.dp(context, if (landscape) 20 else 24)
        val padV = LauncherTheme.dp(context, if (landscape) 16 else 28)
        val btnSize = LauncherTheme.dp(context, if (landscape) 42 else 56)
        val btnSpacing = LauncherTheme.dp(context, if (landscape) 4 else 6)
        val sectionGap = LauncherTheme.dp(context, if (landscape) 8 else 16)
        val keypadTop = LauncherTheme.dp(context, if (landscape) 12 else 24)

        val root = LinearLayout(context)
        root.orientation = LinearLayout.VERTICAL
        root.gravity = Gravity.CENTER_HORIZONTAL
        root.setPadding(padH, padV, padH, padV)
        root.setBackgroundResource(R.drawable.launcher_dialog_bg)

        // 标题
        val title = TextView(context)
        title.setText(if (mode == MODE_SET) R.string.game_password_set_title else R.string.game_password_enter_title)
        title.gravity = Gravity.CENTER
        title.setTextColor(LauncherTheme.text(context))
        title.setTextSize(16f)
        title.setTypeface(null, Typeface.BOLD)
        root.addView(title)

        // 提示文字（初始为空，仅用于动态消息：确认提示、错误提示）
        val hint = TextView(context)
        hint.setText("")
        hint.gravity = Gravity.CENTER
        hint.setTextColor(LauncherTheme.textMuted(context))
        hint.setTextSize(11f)
        val hintLp = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        hintLp.topMargin = LauncherTheme.dp(context, if (landscape) 8 else 12)
        root.addView(hint, hintLp)

        // 密码圆点指示器
        val dots = arrayOfNulls<View>(PASSWORD_LENGTH)
        val dotsRow = LinearLayout(context)
        dotsRow.orientation = LinearLayout.HORIZONTAL
        dotsRow.gravity = Gravity.CENTER
        val dotsRowLp = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        dotsRowLp.topMargin = sectionGap
        val dotSize = LauncherTheme.dp(context, 12)
        val dotSpacing = LauncherTheme.dp(context, 10)
        for (i in 0 until PASSWORD_LENGTH) {
            val dot = View(context)
            val dotBg = GradientDrawable()
            dotBg.shape = GradientDrawable.OVAL
            dotBg.setColor(LauncherTheme.bg(context))
            dot.background = dotBg
            val dotLp = LinearLayout.LayoutParams(dotSize, dotSize)
            if (i > 0) dotLp.leftMargin = dotSpacing
            dotsRow.addView(dot, dotLp)
            dots[i] = dot
        }
        root.addView(dotsRow, dotsRowLp)

        // 九宫格数字键盘
        val input = StringBuilder()
        var firstInputDone = false
        var firstInput = ""

        val keypad = LinearLayout(context)
        keypad.orientation = LinearLayout.VERTICAL
        keypad.gravity = Gravity.CENTER_HORIZONTAL
        val keypadLp = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        keypadLp.topMargin = keypadTop

        val keys = arrayOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "", "0", "del")

        /** 处理一次完整输入：设置模式两次确认、验证模式比对哈希。状态经闭包变量读写保持语义一致。 */
        fun handleComplete(context: Context, dialog: Dialog, mode: Int, entered: String,
                           hashedPassword: String?, hint: TextView,
                           setListener: OnPasswordSetListener?, verifySuccess: Runnable?) {
            if (mode == MODE_SET) {
                if (!firstInputDone) {
                    firstInput = entered
                    firstInputDone = true
                    hint.setText(R.string.game_password_confirm_again)
                } else {
                    if (entered == firstInput) {
                        try {
                            val hashed = hash(entered)
                            dialog.dismiss()
                            setListener?.onPasswordSet(hashed)
                        } catch (e: IllegalStateException) {
                            hint.setText(R.string.game_password_encrypt_save_failed)
                            shakeError(hint)
                            Toast.makeText(context, R.string.game_password_encrypt_failed, Toast.LENGTH_SHORT).show()
                        }
                    } else {
                        firstInputDone = false
                        firstInput = ""
                        hint.setText(R.string.game_password_mismatch)
                        shakeError(hint)
                    }
                }
            } else {
                try {
                    val hashed = hash(entered)
                    if (hashed == hashedPassword) {
                        dialog.dismiss()
                        verifySuccess?.run()
                    } else {
                        hint.setText(R.string.game_password_wrong_retry)
                        shakeError(hint)
                        Toast.makeText(context, R.string.game_password_wrong, Toast.LENGTH_SHORT).show()
                    }
                } catch (e: IllegalStateException) {
                    hint.setText(R.string.game_password_encrypt_verify_failed)
                    shakeError(hint)
                    Toast.makeText(context, R.string.game_password_encrypt_failed, Toast.LENGTH_SHORT).show()
                }
            }
        }

        for (row in 0 until 4) {
            val rowLayout = LinearLayout(context)
            rowLayout.orientation = LinearLayout.HORIZONTAL
            rowLayout.gravity = Gravity.CENTER
            for (col in 0 until 3) {
                val idx = row * 3 + col
                val key = keys[idx]
                if (key.isEmpty()) {
                    val spacer = View(context)
                    val spacerLp = LinearLayout.LayoutParams(btnSize, btnSize)
                    if (col > 0) spacerLp.leftMargin = btnSpacing
                    rowLayout.addView(spacer, spacerLp)
                    continue
                }
                val btn = TextView(context)
                btn.setText(if (key == "del") "X" else key)
                btn.gravity = Gravity.CENTER
                btn.setTextSize(20f)
                btn.setTypeface(null, Typeface.NORMAL)
                btn.setTextColor(LauncherTheme.primary(context))
                val btnLp = LinearLayout.LayoutParams(btnSize, btnSize)
                if (col > 0) btnLp.leftMargin = btnSpacing
                rowLayout.addView(btn, btnLp)

                btn.setOnClickListener { v ->
                    v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
                    if (key == "del") {
                        if (input.isNotEmpty()) {
                            input.deleteCharAt(input.length - 1)
                        }
                    } else {
                        if (input.length < PASSWORD_LENGTH) {
                            input.append(key)
                        }
                    }
                    updateDots(dots, input.length, context)

                    if (input.length == PASSWORD_LENGTH) {
                        val entered = input.toString()
                        input.setLength(0)
                        // 延迟清空圆点，让用户看到最后一个点亮起
                        dots[0]?.postDelayed({
                            if (!dialog.isShowing) return@postDelayed
                            updateDots(dots, 0, context)
                            handleComplete(context, dialog, mode, entered, hashedPassword, hint, setListener, verifySuccess)
                        }, 100)
                    }
                }
            }
            val rowLp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
            if (row > 0) rowLp.topMargin = btnSpacing
            keypad.addView(rowLayout, rowLp)
        }
        root.addView(keypad, keypadLp)

        // 取消按钮
        val cancelBtn = TextView(context)
        cancelBtn.setText(R.string.game_common_cancel)
        cancelBtn.gravity = Gravity.CENTER
        cancelBtn.setTextSize(13f)
        cancelBtn.setTypeface(null, Typeface.BOLD)
        cancelBtn.setTextColor(LauncherTheme.primary(context))
        val cancelLp = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT, LauncherTheme.dp(context, 38))
        cancelLp.topMargin = LauncherTheme.dp(context, if (landscape) 12 else 20)
        cancelBtn.background = LauncherTheme.cancelChip(context)
        cancelBtn.setOnClickListener { dialog.dismiss() }
        root.addView(cancelBtn, cancelLp)

        val scroll = ScrollView(context)
        scroll.addView(root)
        dialog.setContentView(scroll)
        dialog.show()
        LauncherMotion.applyDialogMotion(dialog)
        val window = dialog.window
        if (window != null) {
            window.setBackgroundDrawableResource(android.R.color.transparent)
            val dialogHeight: Int
            if (landscape) {
                val screenHeight = context.resources.displayMetrics.heightPixels
                dialogHeight = (screenHeight * 0.9f).toInt()
            } else {
                dialogHeight = WindowManager.LayoutParams.WRAP_CONTENT
            }
            window.setLayout(LauncherDialogFactory.dialogWidthPx(context, 280), dialogHeight)
        }
    }

    private fun updateDots(dots: Array<View?>, count: Int, context: Context) {
        val activeColor = LauncherTheme.primary(context)
        val inactiveColor = LauncherTheme.bg(context)
        for (i in dots.indices) {
            val bg = GradientDrawable()
            bg.shape = GradientDrawable.OVAL
            bg.setColor(if (i < count) activeColor else inactiveColor)
            dots[i]?.background = bg
        }
    }

    private fun shakeError(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
    }
}
